File: static/gallery_res/demo15/demo15.py
"""
[P_1  0
 0  P_3][  L              0
    P_3^TCP_2^TU^(-1)   -Lbar][U  L^(-1)P_1^TB
                               0     `U_bar`][P_2   0
                                               0   I_4]
where

P_i: ℝ^(4×4) 
B: ℝ^(4×4) 
C: ℝ^(4×4) 
L: ℝ^(4×4) 
Lbar: ℝ^(4×4) 
U: ℝ^(4×4) 
`U_bar`: ℝ^(4×4)
"""
import numpy as np
import scipy
import scipy.linalg
from scipy import sparse
from scipy.integrate import quad
from scipy.optimize import minimize


def demo15(P, B, C, L, Lbar, U, U_bar):
    """
    :param :P : ℝ^(4×4)
    :param :B : ℝ^(4×4)
    :param :C : ℝ^(4×4)
    :param :L : ℝ^(4×4)
    :param :Lbar : ℝ^(4×4)
    :param :U : ℝ^(4×4)
    :param :U_bar : ℝ^(4×4)
    """
    P = np.asarray(P, dtype=np.float64)
    B = np.asarray(B, dtype=np.float64)
    C = np.asarray(C, dtype=np.float64)
    L = np.asarray(L, dtype=np.float64)
    Lbar = np.asarray(Lbar, dtype=np.float64)
    U = np.asarray(U, dtype=np.float64)
    U_bar = np.asarray(U_bar, dtype=np.float64)

    _dim_0 = P.shape[0]
    assert P.shape == (_dim_0, 4, 4)
    assert B.shape == (4, 4)
    assert C.shape == (4, 4)
    assert L.shape == (4, 4)
    assert Lbar.shape == (4, 4)
    assert U.shape == (4, 4)
    assert U_bar.shape == (4, 4)

    _ret_0 = np.block([[P[1-1], np.zeros((4, 4))], [np.zeros((4, 4)), P[3-1]]])
    _ret_1 = np.block([[L, np.zeros((4, 4))], [P[3-1].T @ C @ P[2-1].T @ np.linalg.inv(U), -Lbar]])
    _ret_2 = np.block([[U, np.linalg.inv(L) @ P[1-1].T @ B], [np.zeros((4, 4)), U_bar]])
    _ret_3 = np.block([[P[2-1], np.zeros((4, 4))], [np.zeros((4, 4)), np.identity(4)]])
    ret = _ret_0 @ _ret_1 @ _ret_2 @ _ret_3
    return ret


def generateRandomData():
    _dim_0 = np.random.randint(10)
    P = np.random.randn(_dim_0, 4, 4)
    B = np.random.randn(4, 4)
    C = np.random.randn(4, 4)
    L = np.random.randn(4, 4)
    Lbar = np.random.randn(4, 4)
    U = np.random.randn(4, 4)
    U_bar = np.random.randn(4, 4)
    return P, B, C, L, Lbar, U, U_bar


if __name__ == '__main__':
    P, B, C, L, Lbar, U, U_bar = generateRandomData()
    print("P:", P)
    print("B:", B)
    print("C:", C)
    print("L:", L)
    print("Lbar:", Lbar)
    print("U:", U)
    print("U_bar:", U_bar)
    func_value = demo15(P, B, C, L, Lbar, U, U_bar)
    print("func_value: ", func_value)

File: static/gallery_res/demo15/demo15.cpp
/*
[P₁  0
 0   P₃][    L        0
         P₃ᵀCP₂ᵀU⁻¹  -L̃][U  L⁻¹P₁ᵀB
                         0     Ũ   ][P₂   0
                                     0   I_4]

where

P_i: ℝ^(4×4) 
B: ℝ^(4×4) 
C: ℝ^(4×4) 
L: ℝ^(4×4) 
L̃: ℝ^(4×4) 
U: ℝ^(4×4) 
Ũ: ℝ^(4×4)
*/
#include <Eigen/Core>
#include <Eigen/Dense>
#include <Eigen/Sparse>
#include <iostream>
#include <set>

/**
 * demo15
 *
 * @param P  ℝ^(4×4)
 * @param B  ℝ^(4×4)
 * @param C  ℝ^(4×4)
 * @param L  ℝ^(4×4)
 * @param L̃  ℝ^(4×4)
 * @param U  ℝ^(4×4)
 * @param Ũ  ℝ^(4×4)
 * @return ret
 */
Eigen::Matrix<double, 8, 8> demo15(
    const std::vector<Eigen::Matrix<double, 4, 4>> & P,
    const Eigen::Matrix<double, 4, 4> & B,
    const Eigen::Matrix<double, 4, 4> & C,
    const Eigen::Matrix<double, 4, 4> & L,
    const Eigen::Matrix<double, 4, 4> & L̃,
    const Eigen::Matrix<double, 4, 4> & U,
    const Eigen::Matrix<double, 4, 4> & Ũ)
{
    const long _dim_0 = P.size();
    assert( P.size() == _dim_0 );

    Eigen::Matrix<double, 8, 8> _ret_0;
    _ret_0 << P.at(1-1), Eigen::MatrixXd::Zero(4, 4),
    Eigen::MatrixXd::Zero(4, 4), P.at(3-1);
    Eigen::Matrix<double, 8, 8> _ret_1;
    _ret_1 << L, Eigen::MatrixXd::Zero(4, 4),
    P.at(3-1).transpose() * C * P.at(2-1).transpose() * U.inverse(), -L̃;
    Eigen::Matrix<double, 8, 8> _ret_2;
    _ret_2 << U, L.inverse() * P.at(1-1).transpose() * B,
    Eigen::MatrixXd::Zero(4, 4), Ũ;
    Eigen::Matrix<double, 8, 8> _ret_3;
    _ret_3 << P.at(2-1), Eigen::MatrixXd::Zero(4, 4),
    Eigen::MatrixXd::Zero(4, 4), Eigen::MatrixXd::Identity(4, 4);
    Eigen::Matrix<double, 8, 8> ret = _ret_0 * _ret_1 * _ret_2 * _ret_3;
    return ret;
}


void generateRandomData(std::vector<Eigen::Matrix<double, 4, 4>> & P,
    Eigen::Matrix<double, 4, 4> & B,
    Eigen::Matrix<double, 4, 4> & C,
    Eigen::Matrix<double, 4, 4> & L,
    Eigen::Matrix<double, 4, 4> & L̃,
    Eigen::Matrix<double, 4, 4> & U,
    Eigen::Matrix<double, 4, 4> & Ũ)
{
    const int _dim_0 = rand()%10;
    P.resize(_dim_0);
    for(int i=0; i<_dim_0; i++){
        P[i] = Eigen::MatrixXd::Random(4, 4);
    }
    B = Eigen::MatrixXd::Random(4, 4);
    C = Eigen::MatrixXd::Random(4, 4);
    L = Eigen::MatrixXd::Random(4, 4);
    L̃ = Eigen::MatrixXd::Random(4, 4);
    U = Eigen::MatrixXd::Random(4, 4);
    Ũ = Eigen::MatrixXd::Random(4, 4);
}


int main(int argc, char *argv[])
{
    srand((int)time(NULL));
    std::vector<Eigen::Matrix<double, 4, 4>> P;
    Eigen::Matrix<double, 4, 4> B;
    Eigen::Matrix<double, 4, 4> C;
    Eigen::Matrix<double, 4, 4> L;
    Eigen::Matrix<double, 4, 4> L̃;
    Eigen::Matrix<double, 4, 4> U;
    Eigen::Matrix<double, 4, 4> Ũ;
    generateRandomData(P, B, C, L, L̃, U, Ũ);
    Eigen::Matrix<double, 8, 8> func_value = demo15(P, B, C, L, L̃, U, Ũ);
    std::cout<<"func_value:\n"<<func_value<<std::endl;
    return 0;
}